networkx/algorithms/community/modularity_max.py

```diff
@@ -234,7 +220,115 @@ def greedy_modularity_communities(G, weight=None, resolution=1, n_communities=1)
             b[v] += b[u]
             b[u] = 0
 
-    return sorted(communities.values(), key=len, reverse=True)
+        yield sorted(communities.values(), key=len, reverse=True)
+
+
+def greedy_modularity_communities(
+    G, weight=None, resolution=1, cutoff=None, best_n_communities=None, **aliases
+):
+
+    """
+    Parameters
+    ----------
+    G : NetworkX graph
+
+    weight : string or None, optional (default=None)
+        The name of an edge attribute that holds the numerical value used
+        as a weight.  If None, then each edge has weight 1.
+        The degree is the sum of the edge weights adjacent to the node.
+
+    resolution : float, optional (default=1)
+        If resolution is less than 1, modularity favors larger communities.
+        Greater than 1 favors smaller communities.
+
+    cutoff : int or None, optional (default=None)
+        stop iterating when reaching ``cutoff`` community size: the community merging
+        process is terminated once this number of communities is reached, or until
+        modularity can not be further increased. Must be between 1 and the
+        total number of nodes in `G`. Default is ``None``, meaning the community
+        merging process continues until all nodes are either in (a) the same community
+        or (b) until the best community structure is found.
+
+    best_n_communities : int or None, optional (default=None)
+        the merging proceeds continues until the number of communities
+        ``best_n_communities`` is reached, even if this does not further improve the
+        community structure. In this case those nodes are aggregated that decrease
+        modularity the least.
+
+    Returns
+    -------
+    partition: list
+        A list of frozensets of nodes, one for each community.
+        Sorted by length with largest communities first.
+    """
+    if cutoff is not None:
+        if (cutoff < 1) or (cutoff > G.number_of_nodes()):
+            raise ValueError(
+                f"cutoff must be between 1 and {G.number_of_nodes()}. Got {cutoff}"
+            )
+    if best_n_communities is not None:
+        if (best_n_communities < 1) or (best_n_communities > G.number_of_nodes()):
+            raise ValueError(
+                f"best_n_communities must be between 1 and {G.number_of_nodes()}. "
+                f"Got {best_n_communities}"
+            )
+
+    if "n_communities" in aliases:
```

Comment:
Hmmmm, the `n_communities` kwarg was always integer, never none. Default setting was `n_communities=1` because that kept the original functionality of the function. So only ways to check if `n_communities` was explicitly set by a user would be using the ** option, or not? If not specified explicitly, one wouldn't notice if the default setting is `n_communities=1` or `cutoff=None`. Not sure though.

Reply:
I agree.... we should not check if `n_communities is not None`. Sorry about that.
But setting cutoff to 1 is the same as setting cutoff to None. Right?  The process continues until there is only 1 community in either case. So, maybe we just make the default values `cutoff=1` and `n_communities=1` and check if they are not 1 instead of not None.  ???

Something like: (after changing default `cutoff` value to 1)
```python
if n_communities != 1:
    if cutoff == 1:
        cutoff = n_communities
    else:
        raise error # both cannot be set
```
Does this cover all cases?

Reply:
I partly agree with setting `cutoff` to default value 1 because on the one hand it's somewhat cleaner, but on the other hand that's inconsistent with `best_n=None` (should be `G.number_of_nodes()` by default, but this doesn't work).

For the latter; if `n_communities` is not given, your suggestion currently would result in an error. We could specify it in the kwargs `n_communities=1` as default (as it is currently), but then I don't know how then to distinguish if it was set by a user or if it just the default setting of the function. I mean, a user could also have specified `n_communities=1`, even though that's default. But someone who does not use `n_communities` should not be punished with a deprecation warning, is my feeling...

Reply:
I agree with your default value of `n_communities=None`. I had forgotten about the deprecation tests already. :{
I think we should use default values of `cutoff=1, best_n=None`.  They should not be the same type of default because `best_n` indicates two pieces of information: whether to continue until there are n communities (None vs a number) *and* the value of *n* (the number -- but only if we using this stopping method).  

I think we should have keywords: `n_communities=None, cutoff=1, best_n=None`. We can then check setting of  `n_communities` by checking whether it is `None` or not, `cutoff` works without preprocessing once deprecation is handled, and `best_n` hold two pieces of information in a single keyword. 

I think this means your fix for my suggested code is:
```python
if n_communities is not None:
    create a deprecation warning
    if cutoff == 1:
        cutoff = n_communities
    else:
        raise error # both cannot be set
```
Am I understanding correctly?

Reply:
Another issue to deal with:  For the case when `best_n` is a smaller number than `cutoff`, what do we do?

`cutoff` means we stop. But `best_n` says we have to continue.  I guess we should raise an error in this case too.

Reply:
I think this can not catch the case if `cutoff=1` is set by a user (ie _not_ the default setting), while they set `n_communities=N` to some (other) number N. In this case `n_communities` is even the stronger kwarg than `cutoff`, which it shouldn't be as it is to be deprecated. So, then my suggestion are to either
(i) keep the default setting `cutoff=None` _and_ change the default setting `n_communities=None` (before was 1) - which will make the code more verboose, or
(ii) change default settings to `cutoff=1` _and_ remove `n_communities` from kwargs in favor of **aliases.

Reply:
Agree with your 2nd comment, will raise an error in this case.

Reply:
Can we replace the `aliases` keyword argument with arguments `n_communities` (in the same position it was in before) and `cutoff`?

Reply:
done. can we resolve this, or should we wait for additional feedback?